Useful functions for telomare-evaluare

hhefesto opened the pull request.

hhefesto committed:
Cofree UPT with evaluation to IExpr on every node

hhefesto committed:
Implement new show MultiLineShowUPT: a UPT multiline indented version of regular show.

hhefesto committed:
Improve name for parseSingleExpresion and cleanup

hhefesto committed:
PrettierIExpr Show instance with indentation

hhefesto committed:
Implement PrettierIExpr Show with pretty `(_,_)`

Instead of prefix: `P _ _`

hhefesto committed:
Evaluate from String to IExpr

hhefesto committed:
Tag IExpr with IO-decoupled rEval

hhefesto committed:
Revert pairs shown as `(_,_)`

The prefix notation is ideal to identify nodes of the AST, and
telomare-evaluare benefits from that.

hhefesto committed:
Fix prelude not being part of the UPT

hhefesto committed:
Add better parsing function for telomare-evaluare

Now in telomare-evaluare you'll be able to either eval a single
expression or put several definitions and use the `main` assignment as
the final expression to evaluate.

hhefesto committed:
Add new telomare-evaluare parse function to Eval

hhefesto committed:
Clean up

sfultong reviewed (commented):
a few requests for modifications

hhefesto commented on src/Telomare.hs:
```diff
@@ -713,6 +715,58 @@ instance Show PrettyIExpr where
     Zero -> "0"
     x -> show x
 
+indentSansFirstLine :: Int -> String -> String
```
I now remember why I hadn't put it there:

```
Module imports form a cycle:
        module ‘Telomare’ (src/Telomare.hs)
        imports module ‘PrettyPrint’ (src/PrettyPrint.hs)
  which imports module ‘Telomare’ (src/Telomare.hs)
```

Do you know how to workaround this?

hhefesto replied:
As you suggested, the workaround was to simply pass all the code using the pretty-print code to the PrettyPrint module

hhefesto commented on src/Telomare/Parser.hs:
```diff
@@ -85,19 +83,70 @@ instance Show PrettyPattern where
     (PrettyPattern (PatternPair x y)) -> "(" <> (show . PrettyPattern $ x) <> ", " <> (show . PrettyPattern $ y) <> ")"
     (PrettyPattern PatternIgnore) -> "_"
 
+newtype MultiLineShowUPT = MultiLineShowUPT UnprocessedParsedTerm
```
Similar error here:

```haskell
Module imports form a cycle:
        module ‘Telomare.TypeChecker’ (src/Telomare/TypeChecker.hs)
        imports module ‘PrettyPrint’ (src/PrettyPrint.hs)
  which imports module ‘Telomare.Parser’ (src/Telomare/Parser.hs)
  which imports module ‘Telomare.TypeChecker’ (src/Telomare/TypeChecker.hs)
```

hhefesto replied:
idem

hhefesto commented on src/Telomare/RunTime.hs:
```diff
@@ -90,6 +90,34 @@ nEval (NExprs m) =
     _        -> throwError $ GenericRunTimeError "nEval: no root frag" Zero
 
+
+-- rEval' :: IExpr -- ^ The enviroment.
```
Removed. This slipped my cleanup for some reason.

hhefesto committed:
Additional cleanup

hhefesto committed:
Move pretty-print code to PrettyPrint module

hhefesto committed:
Move Pretty_ types from Parser into PrettyPrint.hs

To do so, some base types had to be moved to the Telomare module, which
was its original home.
